### PR TITLE
feat(rbac): support for adding conditional permissions

### PR DESCRIPTION
**Pull request by divyanshiGupta**
**Story**: https://issues.redhat.com/browse/RHIDP-1832

This PR adds support for adding conditional permissions via create role form. 

https://github.com/janus-idp/backstage-plugins/assets/20724543/342fa091-0481-49e5-a358-3a68607af13d

### CODE REVIEW

**Comment by debsmita1** [comment]
- The Figma shows the `Rule` dropdown options with descriptions

Screenshot 2024-04-30 at 8 03 56 PM



Screenshot 2024-04-30 at 7 47 41 PM





- Selected condition is not highlighted

Screenshot 2024-04-30 at 8 02 09 PM

Figma

Screenshot 2024-04-30 at 8 02 42 PM






- Number of selected conditional accesses are not shown

Figma

Screenshot 2024-04-30 at 8 13 11 PM




- `Review and create` step showing 0 permission policies when conditional accesses are selected


Screenshot 2024-04-30 at 8 18 30 PM

- Can't fetch permission policies for roles created with conditional policies
Screenshot 2024-04-30 at 8 20 54 PM



- Aren't we showing added conditional policies in the Review step ?

**Comment by divyanshiGupta** [comment]
> * Fix unique key warnings for ConditionsFormRow.tsx:107
> * The Figma shows the `Rule` dropdown options with descriptions
> 
> Screenshot 2024-04-30 at 8 03 56 PM
> Screenshot 2024-04-30 at 7 47 41 PM
> 
> * Selected condition is not highlighted
> 
> Screenshot 2024-04-30 at 8 02 09 PM
> 
> Figma
> 
> Screenshot 2024-04-30 at 8 02 42 PM
> * Number of selected conditional accesses are not shown
> 
> Figma
> 
> Screenshot 2024-04-30 at 8 13 11 PM
> * `Review and create` step showing 0 permission policies when conditional accesses are selected
> 
> Screenshot 2024-04-30 at 8 18 30 PM
> 
> * Can't fetch permission policies for roles created with conditional policies
>   Screenshot 2024-04-30 at 8 20 54 PM
> * Aren't we showing added conditional policies in the Review step ?

@debsmita1 thanks for the input. I was working on fixing these after raising the PR. Most of these should be fixed now.

Also currently we are not showing conditions in overview page as we dont have the UX for that but will show the empty state in case no simple permissions are available.

**Comment by divyanshiGupta** [comment]
cc: @ShiranHi

**Comment by ShiranHi** [comment]
> We didnt have UX for review step but have added a section for conditions and confirmed with Shiran. Also currently we are not showing conditions in overview page as we dont have the UX for that but will show the empty state in case no simple permissions are available.

That is correct. we didn't want to show too much details on the Review step. About the Overview page, we can add this in the future.

Some questions:
1. Can the "Configure access" appear after users selected the plugin (first dropdown) and not after they select the resource type?
2. Can the segmented button text be in small letters like in the [design](https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=3287%3A37049&mode=design&t=RzyRUn37iNMuuaQk-1)? It feels hard to read it when all caps. Also, we should have a background color to the selected button. 
image
3. The minus icon is not aligned with the field, and the "+ADD RULE" should be "+ Add rule" and is too close to the field
image01
4. Also here the minus icon is not aligned with the field, and we should show description of each rule
image02

**Comment by divyanshiGupta** [comment]
> Can the "Configure access" appear after users selected the plugin (first dropdown) and not after they select the resource type?

No, because the rules depends upon the resource type and we show the button only when rules exist for that particular resource type. We had a conversation regarding this before as well.

> Can the segmented button text be in small letters like in the [design](https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=3287%3A37049&mode=design&t=RzyRUn37iNMuuaQk-1)? It feels hard to read it when all caps. 

So we are using material-ui components and there have certain limitations. In case of buttons it capitalises the button text interanlly and there is not straightforward of changing that behaviour. If you go through RHDH all the buttons have capitalised text. This is what I am using https://mui.com/material-ui/react-button-group/.

> Also, we should have a background color to the selected button. 

This is another limitation but might be achieved by some CSS hack.

**Comment by AndrienkoAleksandr** [comment]
@divyanshiGupta, you've done a brilliant work. I've tested your pull request and have some feedback:
- when a role is removed, the corresponding condition policies should also be removed.
- there's currently no way to select a previously created condition and edit/delete it from the UI.

**Comment by divyanshiGupta** [comment]
> @divyanshiGupta, you've done a brilliant work. I've tested your pull request and have some feedback:
> 
> * when a role is removed, the corresponding condition policies should also be removed.
> * there's currently no way to select a previously created condition and edit/delete it from the UI.

@AndrienkoAleksandr Thanks for the feedback.

These points will be covered in this story https://issues.redhat.com/browse/RHIDP-1878.

**Comment by debsmita1** [comment]
A few more observations:
-  fix `1 rules` to `1 rule`
Screenshot 2024-05-06 at 4 00 53 PM
- Should the conditional access count in the policies section match the conditional permission policies count in the review section
Screenshot 2024-05-06 at 4 06 15 PM
Screenshot 2024-05-06 at 4 06 05 PM

- After selecting 'Remove All,' the side panel, upon closure, displays '1 rules' alongside the option to 'Configure Access'

**Comment by divyanshiGupta** [comment]
> A few more observations:
> 
> * fix `1 rules` to `1 rule`
>   Screenshot 2024-05-06 at 4 00 53 PM
> * Should the conditional access count in the policies section match the conditional permission policies count in the review section
>   Screenshot 2024-05-06 at 4 06 15 PM
>   Screenshot 2024-05-06 at 4 06 05 PM
> * After selecting 'Remove All,' the side panel, upon closure, displays '1 rules' alongside the option to 'Configure Access'

@debsmita1 so only one conditional permission policy is being created with n number of rules for each row so for now I am displaying only the number of conditional permission policies.

On remove all , the data from the sidebar is removed so that the user can add new data according to them and then save it again.

**Comment by divyanshiGupta** [comment]
@ShiranHi On remove all, should we remove all of the saved data or just remove it from the sidebar so that the user can add new conditions and save the new data to overwrite previous data?

**Comment by ShiranHi** [comment]
> @ShiranHi On remove all, should we remove all of the saved data or just remove it from the sidebar so that the user can add new conditions and save the new data to overwrite previous data?

When you click the "Remove all" button, all the information on the sidebar will be deleted and back to be like the initial state. The sidebar will stay open, so users can add new information and save it again.

**Comment by debsmita1** [comment]
@divyanshiGupta Could you show a proper error message if the creation of conditional permissions fail ?
Screenshot 2024-05-07 at 6 33 35 PM

File: plugins/rbac/src/components/CreateRole/RoleForm.tsx [hunk 1]
```diff
@@ -139,22 +140,45 @@ export const RoleForm = ({
...
+        .filter(m => m);
+
+      if (cpErr.length > 0) {
         throw new Error(
-          `Role was created successfully but unable to add permissions to the role. ${
-            (permissionsRes as unknown as RoleError).error.message
-          }`,
+          `Role was created successfully but unable to add conditions to the role. ${cpErr.join(
+            '\n',
+          )}`,
```
Comment (divyanshiGupta) [inline plugins/rbac/src/components/CreateRole/RoleForm.tsx:178]: @debsmita1 here you can see conditional permissions error handling.

**Pull request merged**